File: src/sys/java/fan/sys/Process.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   5 Nov 06  Brian Frank  Creation
//
package fan.sys;

import java.io.*;
import java.util.Iterator;
import java.util.Map.Entry;

/**
 * Process
 */
public class Process
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Process make()
  {
    return new Process(new List(Sys.StrType), null);
  }

  public static Process make(List command)
  {
    return new Process(command, null);
  }

  public static Process make(List command, File dir)
  {
    return new Process(command, dir);
  }

  private Process(List command, File dir)
  {
    command(command);
    dir(dir);
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public Type typeof()
  {
    return Sys.ProcessType;
  }

//////////////////////////////////////////////////////////////////////////
// Configuration
//////////////////////////////////////////////////////////////////////////

  public List command() { return command; }
  public void command(List v) { this.command = v; }

  public File dir() { return dir; }
  public void dir(File v)
  {
    checkRun();
    if (v != null && (!v.exists() || !v.isDir()))
      throw ArgErr.make("Invalid working directory: " + v);
    this.dir = v;
  }

  public Map env()
  {
    if (env == null)
    {
      env = new Map(Sys.StrType, Sys.StrType);
      Iterator it = new ProcessBuilder().environment().entrySet().iterator();
      while (it.hasNext())
      {
        Entry entry = (Entry)it.next();
        String key = (String)entry.getKey();
        String val = (String)entry.getValue();
        env.set(key, val);
      }
    }
    return env;
  }

  public boolean mergeErr() { return mergeErr; }
  public void mergeErr(boolean v) { checkRun(); mergeErr = v;  }

  public OutStream out() { return out; }
  public void out(OutStream out) { checkRun(); this.out = out; }

  public OutStream err() { return err; }
  public void err(OutStream err) { checkRun(); this.err = err; }

  public InStream in() { return in; }
  public void in(InStream in) { checkRun(); this.in = in;
  }

//////////////////////////////////////////////////////////////////////////
// Lifecycle
//////////////////////////////////////////////////////////////////////////

  public final Process run()
  {
    checkRun();
    try
    {
      // commands
      String[] strings = new String[command.sz()];
      for (int i=0; i<command.sz(); ++i)
        strings[i] = (String)command.get(i);
      ProcessBuilder builder = new ProcessBuilder(strings);

      // environment
      if (env != null)
      {
        Iterator it = env.pairsIterator();
        while (it.hasNext())
        {
          Entry entry = (Entry)it.next();
          String key = (String)entry.getKey();
          String val = (String)entry.getValue();
          builder.environment().put(key, val);
        }
      }

      // working directory
      if (dir != null)
        builder.directory(((LocalFile)dir).file);

      // mergeErr
      if (mergeErr)
        builder.redirectErrorStream(true);

      // map Fantom streams to Java streams


      // start it
      this.proc = builder.start();

      // now launch threads to pipe std out, in, and err
      new PipeInToOut(this, "out", proc.getInputStream(), out).start();
      if (!mergeErr) new PipeInToOut(this, "err", proc.getErrorStream(), err).start();
      if (in != null) new PipeOutToIn(this, proc.getOutputStream(), in).start();

      return this;
    }
    catch (Throwable e)
    {
      this.proc = null;
      throw Err.make(e);
    }
  }

  public final long join()
  {
    if (proc == null) throw Err.make("Process not running");
    try
    {
      return proc.waitFor();
    }
    catch (Throwable e)
    {
      throw Err.make(e);
    }
  }

  public final Process kill()
  {
    if (proc == null) throw Err.make("Process not running");
    proc.destroy();
    return this;
  }

  private void checkRun()
  {
    if (proc != null) throw Err.make("Process already run");
  }

  boolean isAlive()
  {
    // hacky to use exception for flow control, but there
    // doesn't seem to be any other way to check state
    try
    {
      proc.exitValue();
      return false;
    }
    catch (IllegalThreadStateException e)
    {
      return true;
    }
  }

//////////////////////////////////////////////////////////////////////////
// PipeInToOut
//////////////////////////////////////////////////////////////////////////

  static class PipeInToOut extends java.lang.Thread
  {
    PipeInToOut(Process proc, String name, InputStream in, OutStream out)
    {
      super("Process." +  name);
      this.proc = proc;
      this.in   = in;
      this.out  = out == null ? null : SysOutStream.java(out);
    }

    public void run()
    {
      byte[] temp = new byte[256];
      while (true)
      {
        try
        {
          int n = in.read(temp, 0, temp.length);
          if (n < 0) { if (out != null) out.flush(); break; }
          if (out != null) out.write(temp, 0, n);
        }
        catch (Throwable e)
        {
          if (proc.isAlive()) e.printStackTrace();
          else return;
        }
      }
    }

    final Process proc;
    final InputStream in;
    final OutputStream out;
  }

//////////////////////////////////////////////////////////////////////////
// PipeOutToIn
//////////////////////////////////////////////////////////////////////////

  static class PipeOutToIn extends java.lang.Thread
  {
    PipeOutToIn(Process proc, OutputStream out, InStream in)
    {
      super("Process.in");
      this.proc = proc;
      this.out = out;
      this.in  = SysInStream.java(in);
    }

    public void run()
    {
      byte[] temp = new byte[256];
      while (true)
      {
        try
        {
          int n = in.read(temp, 0, temp.length);
          if (n < 0) break;
          out.write(temp, 0, n);
          out.flush();
        }
        catch (Throwable e)
        {
          if (proc.isAlive()) e.printStackTrace();
          else return;
        }
      }
    }

    final Process proc;
    final OutputStream out;
    final InputStream in;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private List command;
  private File dir;
  private Map env;
  private boolean mergeErr = true;
  private OutStream out = Env.cur().out();
  private OutStream err = Env.cur().err();
  private InStream in   = null;
  private volatile java.lang.Process proc;

}

File: src/sys/dotnet/fan/sys/Process.cs
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   6 Feb 08  Andy Frank  Creation
//

using System;
using System.Collections;
using System.Text;

namespace Fan.Sys
{
  /// <summary>
  /// Process
  /// </summary>
  public class Process : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Process make()
    {
      return new Process(new List(Sys.StrType), null);
    }

    public static Process make(List command)
    {
      return new Process(command, null);
    }

    public static Process make(List command, File dir)
    {
      return new Process(command, dir);
    }

    private Process(List command, File dir)
    {
      this.command(command);
      this.dir(dir);
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override Type @typeof()
    {
      return Sys.ProcessType;
    }

  //////////////////////////////////////////////////////////////////////////
  // Configuration
  //////////////////////////////////////////////////////////////////////////

    public List command() { return m_command; }
    public void command(List v) { this.m_command = v; }

    public File dir() { return m_dir; }
    public void dir(File v)
    {
      checkRun();
      if (v != null && (!v.exists() || !v.isDir()))
        throw ArgErr.make("Invalid working directory: " + v).val;
      this.m_dir = v;
    }

    public Map env()
    {
      if (m_env == null)
      {
        m_env = new Map(Sys.StrType, Sys.StrType);
        IDictionaryEnumerator en =
         (IDictionaryEnumerator)new System.Diagnostics.Process().
           StartInfo.EnvironmentVariables.GetEnumerator();
        while (en.MoveNext())
        {
          string key = (string)en.Key;
          string val = (string)en.Value;
          m_env.set(key, val);
        }
      }
      return m_env;
    }

    public bool mergeErr() { return m_mergeErr; }
    public void mergeErr(bool v) { checkRun(); m_mergeErr = v;  }

    public OutStream @out() { return m_out; }
    public void @out(OutStream @out) { checkRun(); this.m_out = @out; }

    public OutStream err() { return m_err; }
    public void err(OutStream err) { checkRun(); this.m_err = err; }

    public InStream @in() { return m_in; }
    public void @in(InStream @in) { checkRun(); this.m_in = @in; }

  //////////////////////////////////////////////////////////////////////////
  // Lifecycle
  //////////////////////////////////////////////////////////////////////////

    public Process run()
    {
      checkRun();
      try
      {
        // arguments
        string fileName = m_command.get(0) as string;
        StringBuilder args = new StringBuilder();
          for (int i=1; i<m_command.sz(); i++)
          {
            if (i > 1) args.Append(" ");
            args.Append(m_command.get(i) as string);
          }

        // create process
        m_proc = new System.Diagnostics.Process();
        m_proc.StartInfo.UseShellExecute = false;
        m_proc.StartInfo.FileName = fileName;
        m_proc.StartInfo.Arguments = args.ToString();

        // environment
        if (m_env != null)
        {
          IDictionaryEnumerator en = m_env.pairsIterator();
          while (en.MoveNext())
          {
            string key = (string)en.Key;
            string val = (string)en.Value;
            m_proc.StartInfo.EnvironmentVariables[key] = val;
          }
        }

        // working directory
        if (m_dir != null)
          m_proc.StartInfo.WorkingDirectory = ((LocalFile)m_dir).m_file.FullName;

        // streams
        if (m_in != null) m_proc.StartInfo.RedirectStandardInput = true;
        m_proc.StartInfo.RedirectStandardOutput = true;
        m_proc.StartInfo.RedirectStandardError  = true;
        m_proc.OutputDataReceived += new System.Diagnostics.DataReceivedEventHandler(outHandler);
        m_proc.ErrorDataReceived  += new System.Diagnostics.DataReceivedEventHandler(errHandler);

        // start it
        m_proc.Start();

        // start async read/writes
        if (m_in != null)
        {
          new System.Threading.Thread(
            new System.Threading.ThreadStart(inHandler)).Start();
        }
        m_proc.BeginOutputReadLine();
        m_proc.BeginErrorReadLine();

        return this;
      }
      catch (System.Exception e)
      {
        m_proc = null;
        throw Err.make(e).val;
      }
    }

    public long join()
    {
      if (m_proc == null) throw Err.make("Process not running").val;
      try
      {
        m_proc.WaitForExit();
        return m_proc.ExitCode;
      }
      catch (System.Exception e)
      {
        throw Err.make(e).val;
      }
    }

    private void checkRun()
    {
      if (m_proc != null) throw Err.make("Process already run").val;
    }

  //////////////////////////////////////////////////////////////////////////
  // Handlers
  //////////////////////////////////////////////////////////////////////////

    private void inHandler()
    {
      System.IO.Stream input  = SysInStream.dotnet(m_in);
      System.IO.Stream output = m_proc.StandardInput.BaseStream;
      byte[] temp = new byte[256];

      while (!m_proc.HasExited)
      {
        try
        {
          int n = input.Read(temp, 0, temp.Length);
          if (n < 0) break;
          output.Write(temp, 0, n);
          output.Flush();
        }
        catch (System.Exception e)
        {
          Err.dumpStack(e);
        }
      }
    }

    private void outHandler(object sender, System.Diagnostics.DataReceivedEventArgs args)
    {
      if (String.IsNullOrEmpty(args.Data)) return;
      if (m_out != null) m_out.printLine(args.Data);
    }

    private void errHandler(object sender, System.Diagnostics.DataReceivedEventArgs args)
    {
      if (String.IsNullOrEmpty(args.Data)) return;
      if (m_mergeErr)
      {
        if (m_out != null) m_out.printLine(args.Data);
      }
      else
      {
        if (m_err != null) m_err.printLine(args.Data);
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    private List m_command;
    private File m_dir;
    private Map m_env;
    private bool m_mergeErr = true;
    private OutStream m_out = Env.cur().@out();
    private OutStream m_err = Env.cur().err();
    private InStream m_in   = null;
    private volatile System.Diagnostics.Process m_proc;

  }
}